x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -0,0 +1,130 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.features.FeatureService;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.UpdateByQueryAction;
+import org.elasticsearch.index.reindex.UpdateByQueryRequest;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptType;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_MIGRATION_FRAMEWORK;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SecurityMainIndexMappingVersion.ADD_REMOTE_CLUSTER_AND_DESCRIPTION_FIELDS;
+
+public class SecurityMigrations {
+
+    public interface SecurityMigration {
+        /**
+         * Method that will execute the actual migration - needs to be idempotent and non-blocking
+         *
+         * @param indexManager for the security index
+         * @param client the index client
+         * @param listener listener to provide updates back to caller
+         */
+        void migrate(SecurityIndexManager indexManager, Client client, ActionListener<Void> listener);
+
+        /**
+         * Any node features that are required for this migration to run. This makes sure that all nodes in the cluster can handle any
+         * changes in behaviour introduced by the migration.
+         *
+         * @return a set of features needed to be supported or an empty set if no change in behaviour is expected
+         */
+        Set<NodeFeature> nodeFeaturesRequired();
+
+        /**
+         * The min mapping version required to support this migration. This makes sure that the index has at least the min mapping that is
+         * required to support the migration.
+         *
+         * @return the minimum mapping version required to apply this migration
+         */
+        int minMappingVersion();
+    }
+
+    public static final TreeMap<Integer, SecurityMigration> MIGRATIONS_BY_VERSION = new TreeMap<>(Map.of(1, new SecurityMigration() {
+        private static final Logger logger = LogManager.getLogger(SecurityMigration.class);
+
+        @Override
+        public void migrate(SecurityIndexManager indexManager, Client client, ActionListener<Void> listener) {
+            BoolQueryBuilder filterQuery = new BoolQueryBuilder().filter(QueryBuilders.termQuery("type", "role"));
```

Comment:
Have you considered also checking that the `metadata_flattened` is non-`null` to do the reindexing. I think the `Preparing to migrate X roles" in the log can be misleading otherwise.

Reply:
Added a check for this. Thanks!